gdb/riscv-tdep.c

```diff
@@ -1301,42 +1241,139 @@ riscv_gdbarch_init (struct gdbarch_info info,
+              named_csr[i].num,
+              {named_csr[i].name, generic_name},
+              false,
+              false,
+              "org.gnu.gdb.riscv.csr",
+              all_reggroup
+          };
+
+          riscv_reg_info.push_back(reg);
+        }
+      reg_info_built = true;
+    }
+
+  bool use_tdesc_registers = false;
   if (tdesc_has_registers (info.target_desc))
     {
-      const struct tdesc_feature *feature;
-      struct tdesc_arch_data *tdesc_data;
-      int valid_p;
+      use_tdesc_registers = true;
 
-      feature = tdesc_find_feature (info.target_desc, "org.gnu.gdb.riscv.cpu");
-      if (feature == NULL)
-	goto no_tdata;
+      struct tdesc_arch_data *tdesc_data = tdesc_data_alloc ();
 
-      tdesc_data = tdesc_data_alloc ();
+      std::map<int, const char*> found;
 
-      valid_p = 1;
-      for (i = RISCV_ZERO_REGNUM; i <= RISCV_LAST_FP_REGNUM; ++i)
-        valid_p &= tdesc_numbered_register (feature, tdesc_data, i,
-                                            riscv_gdb_reg_names[i]);
-      for (i = RISCV_FIRST_CSR_REGNUM; i <= RISCV_LAST_CSR_REGNUM; ++i)
+      for (auto reg_info = riscv_reg_info.begin();
+           reg_info != riscv_reg_info.end(); ++reg_info)
         {
-          char buf[20];
-          sprintf (buf, "csr%d", i - RISCV_FIRST_CSR_REGNUM);
-          valid_p &= tdesc_numbered_register (feature, tdesc_data, i, buf);
+          const struct tdesc_feature *feature =
+            tdesc_find_feature (info.target_desc, reg_info->feature_name);
+          int success = 0;
+          if (feature)
+            // Look for this register by any of its names.
+            for (auto name = reg_info->names.begin();
+                 name != reg_info->names.end(); ++name)
+              {
+                success = tdesc_numbered_register (feature, tdesc_data,
+                                                   reg_info->number, *name);
+                if (success)
+                  {
+                    found[reg_info->number] = *name;
+                    break;
+                  }
+              }
+
+          if (!success && reg_info->required)
+            {
+              use_tdesc_registers = false;
+              break;
+            }
         }
 
-      valid_p &= tdesc_numbered_register (feature, tdesc_data, i++, "priv");
-
-      if (!valid_p)
-	tdesc_data_cleanup (tdesc_data);
+      if (use_tdesc_registers)
+        {
+          // This number apparently must be at least as large as the number of
+          // registers in the target description. (See assertion in
+          // tdesc_use_registers().) But that seems to defeat part of the point
+          // of the target description, so I might be missing something.
+          set_gdbarch_num_regs (gdbarch, RISCV_NUM_REGS);
+          set_gdbarch_sp_regnum (gdbarch, RISCV_SP_REGNUM);
+          set_gdbarch_pc_regnum (gdbarch, RISCV_PC_REGNUM);
+          set_gdbarch_ps_regnum (gdbarch, RISCV_FP_REGNUM);
```

Comment:
Setting the processor status register to point at the frame pointer register looks odd, but it was that way before your patch.

Reply:
The actual frame examining code is voodoo to me, so I just try not to touch it.